### DISCUSSION TITLE
Pico: eliminate unplug and push BOOTSEL to flash

### DISCUSSION THREAD

**Discussion by phoddie** [Moddable-OpenSource/moddable#573] [Porting]
Installing a Moddable SDK project onto the Pico is a little too manual. Each time it is necessary to unplug (or otherwise power cycle) the device while holding down the BOOTSEL button. That's like programming Espressif ESP8266 and ESP32 boards that lack auto-programming circuitry. To streamline builds, I'd like to eliminate the extra step.

Espressif devices use a clever hardware trick both to reboot the device and to enter programming mode. They use the RTS and DTR pins on the serial port to control this. The Pico uses USB to connect to the computer (at least by default) so the physical RTS and DTR pins aren't available. But... they are logically present over the USB CDC connection. Maybe it is possible to emulate the Espressif behavior on the Pico?

The CDC support in TinyUSB invokes an optional `tud_cdc_line_state_cb` callback function whenever the RTS or DTR pins change state. By implementing that function in the Moddable SDK Pico port, we can receive the same RTS and DTR signals as on the Espressif boards and interpret them to both reboot and enter programming mode.

Here's a simple implementation that I added to [debugger.c](https://github.com/Moddable-OpenSource/moddable/blob/pico/build/devices/pico/base/debugger.c).:

```c
void tud_cdc_line_state_cb(uint8_t itf, bool dtr, bool rts)
{
	static int8_t lastRTS = -1;
	if (-1 != lastRTS) {
		if (lastRTS && !rts)
			pico_reboot(dtr);
	}
	lastRTS = rts;
}
```

If you add this to your build, you get a couple of new capabilities. First, you can debug the currently installed JavaScript app on the Pico without having to also install. Here's the command:

```
mcconfig -d -m -p pico -t xsbug
```

Second, you can force the Pico into programming mode without having to hold down the BOOTSEL button while cycling the power. To do that, use the (undocumented) `-programming` option with `serial2xsbug`. You'll need to adjust the serial driver path to match your installation.

```
serial2xsbug /dev/cu.usbmodem0000000000001 921600 8N1 -programming
```

Here's an example. When you see the "RPI-RP2" disk appear on the desktop, the Pico is ready for programming.

![mount](https://user-images.githubusercontent.com/6856458/108609159-f77b6780-7380-11eb-9ac5-e1d24c15512a.gif)

This only works if the Moddable SDK is installed with the `tud_cdc_line_state_cb` function above and a serial connection has been established over USB. If not, you still need to do the power-cycle-with-BOOTSEL-pressed dance.

The next step is to integrate this into the build so that it automatically triggers programming mode before installing the firmware image. This also needs to be tested on Linux and Windows. The serial drivers may behave a little differently there which may require some more complex logic in `tud_cdc_line_state_cb`.

**Comment by mkellner** [comment]
I could use `serial2xsbug 2e8a:000a 921600 8N1 -programming` on my Mac to bypass the serial-port path variability.
`2e8a:000a` is the USB VID/PID (vendor ID/product ID) for the device in runtime.

(to note: `2e8a:0003` is the VID/PID for the Pico in boot mode.)
